CollAction/Frontend/app/global/NewsletterSignup.tsx

```diff
@@ -0,0 +1,141 @@
+import * as React from "react";
+import renderComponentIf from "./renderComponentIf";
+import { library } from "@fortawesome/fontawesome-svg-core";
+import { faEnvelope } from "@fortawesome/free-solid-svg-icons";
+import * as classnames from "classnames";
+
+library.add(faEnvelope);
+
+enum States { VOID, ERROR, READY, LOADING }
+
+interface NewsLetterSignupProps {
+  mailchimpUser: string;
+  mailChimpListId: string;
+}
+
+interface NewsLetterSignupState {
+  subscriptionState: States;
+  email?: string;
+  signupStatusMessage?: string;
+}
+
+class NewsletterSignup extends React.Component<NewsLetterSignupProps, NewsLetterSignupState> {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      email: "",
+      subscriptionState: States.VOID
+    };
+
+    this.handleEmailChange = this.handleEmailChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.getFormContainerClassNames = this.getFormContainerClassNames.bind(this);
+  }
+
+  componentDidMount() {}
+
+  getFormContainerClassNames() {
+    return classnames(
+      "newsletter-signup__container",
+      {
+        "loading": this.state.subscriptionState === States.LOADING,
+        "hidden": this.state.subscriptionState === States.READY,
+      }
+    );
+  }
+
+  getStatusMessageClassNames() {
+    return classnames(
+      "newsletter-signup__message",
+      {
+        "show-error" : this.state.subscriptionState === States.ERROR,
+        "show-success" : this.state.subscriptionState === States.READY
+      }
+    );
+  }
+
+  handleEmailChange(event) {
+    this.setState({
+      email: event.target.value,
+      subscriptionState: States.VOID
+    });
+  }
+
+  isValidEmail() {
+    return /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,3}/.test(this.state.email);
+  }
+
+  handleSubmit(event) {
+    event.preventDefault();
+
+    // disable submit if email is invalid
+    if (!this.isValidEmail()) {
+      this.setState({
+        subscriptionState: States.ERROR,
+        signupStatusMessage: "Please enter a valid email address"
+      });
+
+      return;
+    }
+
+    let formData = new FormData();
+    formData.append("MERGE0", this.state.email);
+    formData.append("u", this.props.mailchimpUser);
+    formData.append("id", this.props.mailChimpListId);
+
+    this.setState({ subscriptionState: States.LOADING });
+    
+    fetch("http://collaction.us14.list-manage.com/subscribe/post", {
+      method: "POST",
+      body: formData,
+    })
+    .then(response => {
+      if (response.ok) {
+        this.showSignupSucces();
+      } else {
+        this.showSignupError();
```

Comment:
It might be the case that the http response actually has the signup error message, that might be useful to show as well (big difference between 'already subscribed', 'we suspect you might be a spammer', etc). Might be a good idea to check if that's the case.

Reply:
Fixed it.